web/src/stream_popover.js

```diff
@@ -207,7 +209,107 @@ function build_stream_popover(opts) {
     });
 }
 
-export function build_move_topic_to_stream_popover(
+async function get_message_placement_from_server(
+    current_stream_id,
+    topic_name,
+    current_message_id,
+) {
+    return new Promise((resolve) => {
+        message_edit.is_message_oldest_or_newest(
+            current_stream_id,
+            topic_name,
+            current_message_id,
+            (is_oldest, is_newest) => {
+                if (is_oldest) {
+                    resolve("first");
+                } else if (is_newest) {
+                    resolve("last");
+                } else {
+                    resolve("intermediate");
+                }
+            },
+        );
+    });
+}
+
+async function get_message_placement_in_conversation(
+    current_stream_id,
+    topic_name,
+    current_message_id,
+) {
+    // First we check if the placement of the message can be determined
+    // in the current message list. This allows us to avoid a server call
+    // in most cases.
+
+    // We check if the view contains the entire conversation of which
+    // the message is a part of. The #design stream at CZO is an example of
+    // such a view.
+    if (current.data.filter.supports_collapsing_recipients()) {
+        // Next we check if we are in a conversation view, examples
+        // of such views are the topics and the DM view. If we are
+        // in a conversation view, we check if the message is the
+        // first or the last message in the current view. If not, we
+        // can conclude that the message is an intermediate message.
+        if (current.data.filter.is_conversation_view()) {
+            if (
+                current.data.fetch_status.has_found_oldest() &&
+                current.data.first().id === current_message_id
+            ) {
+                return "first";
+            } else if (
+                current.data.fetch_status.has_found_newest() &&
+                current.data.last().id === current_message_id
+            ) {
+                return "last";
+            }
+            return "intermediate";
+        }
+
+        // If we are not in a conversation view, but still know
+        // the view contains the entire conversation, we check if
+        // we can find the adjacent messages in the current view
+        // through which we can determine if the message is an
+        // intermediate message or not.
+        const msg_list = current.data.all_messages();
+        let is_next_found = false;
+        let is_prev_found = false;
```

Comment:
I renamed these variables to things like `found_newer_matching_message`, both because we're not looking for immediately prev/next, and because `newer`/`older` is a way we talk about this relationship in the rest of this code path.